In [45]:
import numpy as np
import time as t

Q1 Gradient Descent algorithm :


In [46]:
def gradient_descent(gradient,init_,learn_rate, n_iter=50, tol=1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate*gradient(x)
        if np.all(np.abs(delta) <= tol):
            break
        x += deltastochastic
    return round(x*1000)/1000

#this function is given in the question itself

(a) Use this function to find minima for (i) $x^2 + 3x+4$ and (ii) $x^4 – 3x^2 +2x$

**Solution**: To find minima for a polynomial , we will use the above gradient descent algorithm with gradient at any x be its derivative

(i) For $x^2 + 3x + 4$ , we will use learn_rate = 0.1 , initial value of $x$ = 0 and gradient at any x be $2x + 3$

In [47]:
#here we are passing gradient as an inline function which is the derivative of the function given in the question and we are finding the minima using it
min1 = gradient_descent(gradient=lambda x: 2*x + 3, init_= 0, learn_rate=0.2)
print("Minimum value at x = ",min1)
print("Minimum = ", min1**2 + 3*min1 + 4)

Minimum value at x =  -1.5
Minimum =  1.75



(ii) For $x^4 - 3x^2 + 2x$ , we will use learn_rate = 0.1 , initial value of $x$ = 0 and gradient at any x be $4x^3 - 6x + 2$

In [48]:
#We have done the same thing here also
min_2 = gradient_descent(gradient= lambda x: 4*x**3 - 6*x + 2, init_= 0, learn_rate=0.1)
print("Minimum value will be at x = ", min_2)
print("Minimum value = ", min_2**4 - 3*min_2**2 + 2*min_2)

Minimum value will be at x =  -1.366
Minimum value =  -4.848076206064


(b) For linear regression : y  = ax + b , let's take mean squared error to calculate cost function.

Cost Function = $\frac{1}{n}\sum_{i=1}^{n} (y_i - y_{ipred})^2$
. Here $y_{ipred} = ax_i + b$ (theoretically). In gradient descent algorithm, we need to find the gradients at every x in order to update the value of x. In order to calculate the gradient , we need to find the partial derivative of the cost function wrt to a as well as wrt to b.

$D_a = \frac{ \partial ( Cost Function ) }{\partial a} = \frac{ \partial (  \frac{1}{n}\sum_{i=1}^{n} (y_i - y_{ipred})^2 ) }{\partial a} = \frac{-2}{n} \sum_{i=1}^{n} x_i(y_i - y_{ipred})$

$D_b = \frac{ \partial ( Cost Function ) }{\partial b} = \frac{ \partial (  \frac{1}{n}\sum_{i=1}^{n} (y_i - y_{ipred})^2 ) }{\partial b} = \frac{-2}{n} \sum_{i=1}^{n} (y_i - y_{ipred})$


In [87]:
# Gradient Function where we have put in the above calculated gradient function in the code
def gradient(x, X_here = X, Y_here = y):
    (a,b) = x
    N = float(len(Y_here))
    y_curr = (a * X_here) + b
    grad_a = -(2/N) * sum(X_here * (Y_here - y_curr))  #this gives partial derivative wrt a
    grad_b = -(2/N) * sum(Y_here - y_curr)  ##this gives partial derivative wrt b
    return np.array([grad_a,grad_b])

#Here X_here and Y_here are training examples

(c) Generating artificial data and applying gradient descent algorithm to find optimum parameters {a,b}

(i) Generating random training data

In [88]:
#given in the question
np.random.seed(0)
X = 2.5 * np.random.randn(10000) + 1.5
res = 1.5 * np.random.randn(10000)
y = 2 + 0.3* X + res 
#X,y forms the random dataset

In [89]:
def gradient_descent(gradient,init_,learn_rate, n_iter=50, tol=1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate * gradient(x)
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return np.round(x*1000)/1000


Running the linear regression y = ax + b and applying Gradient Descent







In [90]:
#running gradient descent function to find the optimal value of x
gradient_descent(gradient, init_= (0,0), learn_rate=0.001,n_iter = 10000)

array([0.295, 2.023])

(d) Minibatch Stochastic Gradient Descent:

In [91]:
#minibatch sgd function is deined here
def minibatch_sgd(gradient, init_, learn_rate, n_iter = 50, tol = 1e-06, batch_size = 125):
  x = init_
  for _ in range(n_iter):
    indexes = np.random.choice(np.arange(len(X)) , batch_size, replace = False)       #we have taken random indexes which will make our batches
    X_batch = X[indexes]    # X value of those indexes
    y_batch = y[indexes]    # y value of those indexes
    delta = -learn_rate * gradient(x, X_batch, y_batch)   #calculated delta in the similar way
    if np.all(np.abs(delta) <= tol):
      break
    x+= delta

  return np.round(x*1000)/1000

In [92]:
minibatch_sgd(gradient,init_ = (0,0) , learn_rate= 0.001, n_iter = 10000)

array([0.297, 2.021])

(e) We will first compare time for both gradient descent and minibatch sgd:-

In [74]:
#time taken by GD:
tic_gd = t.process_time()   #initial time
gradient_descent(gradient, init_= (0,0), learn_rate=0.001,n_iter = 10000)
toc_gd = t.process_time()   #Final time

time_taken_gd = toc_gd - tic_gd
print("Time taken by gradient descent is: ", time_taken_gd)


Time taken by gradient descent is:  18.605123625000033


In [75]:
#time taken by minibatch stochastic gradient descent:
tic_sgd = t.process_time()    #initial time
minibatch_sgd(gradient, init_= (0,0), learn_rate=0.001,n_iter = 10000)
toc_sgd = t.process_time()    #Final time

time_taken_sgd = toc_sgd - tic_sgd
print("Time taken by minibatch stochastic gradient descent is: ", time_taken_sgd)

Time taken by minibatch stochastic gradient descent is:  2.9779027830000473


Therefore, time taken by minibatch stochastic gradient descent is much lower than normal gradient descent, SGD took approx **3** secs for 10,000 iterations, whereas GD took **18.6** secs for the same 10,000 iterations

In [80]:
min_time = 1e9
optimal_batch_size = 0
for batch_s in range(25,1000,25):   #loop for batch sizes in multiple of 25
  tic_sgd = t.process_time()
  minibatch_sgd(gradient, init_= (0,0), learn_rate=0.001,n_iter = 10000, batch_size = batch_s)
  toc_sgd = t.process_time()

  time_taken_sgd = toc_sgd - tic_sgd    #time for each batch size 
  if(time_taken_sgd < min_time):      #if this time is less than min_time then
    min_time = time_taken_sgd         #update value of min_time
    optimal_batch_size = batch_s      #update optimal_batch size
  
  # print(batch_s)

print(min_time)
print(optimal_batch_size)

2.5451389610000206
225


Therefore, opmitimal batch size that runs minibatch stochastic gradient descent in minimum time is: **225** and the minimum time used for the same is **2.545** secs for 10,000 iterations if we take batch sizes as multiples of 25, we cn calculate the exact optimal batch size by taking increase by 1

##Q2. Finding probability of:
(i) Someone having both cold and a fever

(ii) Someone who has a cough has a cold.

##Solution:-
(i) $P(\text{cold} \cap \text{fever}) = P(\text{fever}/\text{cold}) \times P(\text{cold})=0.307*0.02=0.00614$

(ii)
\begin{align*}
P(\text{cold}/\text{cough}) &= \frac{P(\text{cold } \cap \text{ cough})}{P(\text{cough})}\\ 
 &= \frac{P(\text{cough}/\text{cold})\cdot P(\text{cold})}{P(\text{cough}/\text{cold})\cdot P(\text{cold})+P(\text{cough}/\text{no cold})\cdot P(\text{no cold})} \dots \textbf{ Eqn. 1}\\ \\
\end{align*}


\begin{align*}
P(\text{lung disease}) &= P(\text{lung disease/smoke})\cdot P(\text{smoke}) + P(\text{lung disease/no smoke})\cdot P(\text{no smoke}) \\ &=0.1009*0.2+0.001*0.8 \\ &= 0.02098 \\ \\
\end{align*}


\begin{align*}
P(\text{no lung disease}) &= 1 - P(\text{lung disease})  \\ &= 1 - 0.02098 \\ &= 0.97902 \\
\end{align*}

\begin{align*}
P(cough/cold) &= P(cough/cold/lung disease) \cdot P(lung disease) + P(cough/cold/no lung disease) \cdot P(no lung disease)\\ &= 0.7525*0.02098+0.505*0.97902 \\ &= 0.01578745+0.4944051 \\ &= 0.51019255 \\
\end{align*}

\begin{align*}
P(cough/nocold) &= P(cough/nocold/lung disease) \cdot P(lung disease) + P(cough/nocold/no lung disease) \cdot P(no lung disease)\\ &= 0.505*0.02098+0.01*0.97902 \\ &= 0.0105949+0.490097902 \\ &= 0.0203851 \\
\end{align*}


Putting these values in Eqn $1$, we get
\begin{align*}
P(cold/cough) &=  \frac{0.51019255*0.02}{0.51019255*0.02+0.0203851*0.98}\\ &=  \frac{0.010203851}{0.010203851 + 0.19977398} = \frac{0.010203851}{0.030181249}\\ \\ &= 0.338085776
\end{align*}

Hence, Required answer is $0.338$ aprroximately.

#Q3. 
MLE for a multinomial Distribution

##Solution:

$P(X_1 = x_1,X_2 = x_2,\dots,X_k = x_k) = \frac{n!}{x_1! x_2!\dots x_k!} p_1^{x_1},p_2^{x_2} \dots,p_k^{x_k},  \qquad{0\leq x_i \leq n} \quad(\text{Eqn } 1)$ 

where $\sum_{i=1}^k x_i = n$

And since all the k outcomes are mutually exclusive, therefore $\sum_{i=1}^k p_i = 1$ (Eqn $2$)

The log-likelihood for the probability mass function -
$$log(P(x_1,x_2,\dots,x_k)) = log (n!) - \sum_{i=1}^k log (x_i!) + \sum_{i=1}^k x_i log (p_i!)$$

Now, maximizing the log-likelihood function -
$$L(p) = log (n!) - \sum_{i=1}^k log (x_i!) + \sum_{i=1}^k x_i log (p_i!) + \lambda (1 - \sum_{i=1}^k p_i)$$


Differentiating $L(p)$ wrt $p_i$, we get
$$\frac{\partial L(p)}{\partial p_i} = \frac{x_i}{p_i} - \lambda$$
Setting this to $0$ gives:
$$\frac{x_i}{p_i} = \lambda$$


$$\implies p_i = \frac{x_i}{\lambda} \qquad{\dots \dots \text{Eqn $3$}}$$

Now, we know from Eqn. 2 that $\sum_{i=1}^k p_i = 1$, therefore $\sum_{i=1}^k \frac{x_i}{\lambda} = 1$

$$ \sum_{i=1}^k x_i = \lambda$$

Therefore, 

$$\lambda = n$$

Hence, Eqn $3$ becomes
$$p_i = \frac{x_i}{n}$$ 
which is the MLE for the parameters of a $k$-sided multinomial distribution.